packages/datadog-shimmer/src/shimmer.js

```diff
@@ -1,85 +1,88 @@
+  'caller',
+  'arguments',
+  'name',
+  'length'
+])
+
 function copyProperties (original, wrapped) {
-  // TODO getPrototypeOf is not fast. Should we instead do this in specific
-  // instrumentations where needed?
-  const proto = Object.getPrototypeOf(original)
-  if (proto !== Function.prototype) {
+  if (original.constructor !== wrapped.constructor) {
+    const proto = Object.getPrototypeOf(original)
     Object.setPrototypeOf(wrapped, proto)
   }
-  const props = Object.getOwnPropertyDescriptors(original)
-  const keys = Reflect.ownKeys(props)
-
-  for (const key of keys) {
-    try {
-      Object.defineProperty(wrapped, key, props[key])
-    } catch (e) {
-      // TODO: figure out how to handle this without a try/catch
+
+  const ownKeys = Reflect.ownKeys(original)
+  if (original.length !== wrapped.length) {
+    Object.defineProperty(wrapped, 'length', { value: original.length, configurable: true })
+  }
+  if (original.name !== wrapped.name) {
+    Object.defineProperty(wrapped, 'name', { value: original.name, configurable: true })
+  }
+  if (ownKeys.length !== 2) {
+    for (const key of ownKeys) {
+      if (skipMethods.has(key)) continue
+      const descriptor = Object.getOwnPropertyDescriptor(original, key)
+      if (descriptor.writable || descriptor.configurable || !Object.hasOwn(wrapped, key)) {
+        Object.defineProperty(wrapped, key, descriptor)
+      }
     }
   }
 }
 
 function wrapFunction (original, wrapper) {
-  if (typeof original === 'function') assertNotClass(original)
-
   const wrapped = wrapper(original)
 
-  if (typeof original === 'function') copyProperties(original, wrapped)
+  if (typeof original === 'function') {
+    assertNotClass(original)
+    copyProperties(original, wrapped)
+  }
 
   return wrapped
 }
 
-const wrapFn = function (original, delegate) {
-  throw new Error('calling `wrap()` with 2 args is deprecated. Use wrapFunction instead.')
-}
-
-function wrapMethod (target, name, wrapper, noAssert) {
+function wrap (target, name, wrapper, noAssert) {
   if (!noAssert) {
     assertMethod(target, name)
-    assertFunction(wrapper)
+    if (typeof wrapper !== 'function') {
+      throw new Error(wrapper ? 'Target is not a function' : 'No function provided')
+    }
   }
 
   const original = target[name]
   const wrapped = wrapper(original)
 
-  const descriptor = Object.getOwnPropertyDescriptor(target, name)
-
-  const attributes = {
-    configurable: true,
-    ...descriptor
-  }
-
   if (typeof original === 'function') copyProperties(original, wrapped)
 
+  const descriptor = Object.getOwnPropertyDescriptor(target, name)
+
   if (descriptor) {
     if (descriptor.get || descriptor.set) {
-      attributes.get = () => wrapped
+      // TODO(BridgeAR): What happens in case there is a setter? This seems wrong?
+      // What happens in case the user does indeed set this to a different value?
+      // In that case the getter would potentially return the wrong value?
```

Comment:
Any getter at all is sufficiently a problem here, since a setter doesn't need to exist for its return value to be different each time.

A bailout here would save us from causing damage, but I also think there's a correct approach here where we ensure that the wrapper always calls the original getter on every invocation.

That can be done in a separate PR, since that's a pre-existing issue here. One could also argue that because tests are passing, the bug/shortcut is fine. I'd disagree with that argument, because we don't exercise in tests the full capabilities of libraries we instrument.

Reply:
I would like to address it in a separate PR while agreeing to your perspective.